Sources/NIOHTTP2/HTTP2FrameParser.swift

```diff
@@ -0,0 +1,738 @@
+            let headers: HPACKHeaders
+            if bytes.readableBytes == headerByteLen {
+                headers = try stream.decoder.decodeHeaders(from: &bytes)
+            } else {
+                // slice out the relevant chunk of data
+                var slice = bytes.readSlice(length: headerByteLen)!
+                headers = try stream.decoder.decodeHeaders(from: &slice)
+            }
+            
+            payload = .pushPromise(promisedStreamID, headers)
+            
+        case 6:
+            // PING frame : RFC 7540 § 6.7
+            guard header.length == 8 else {
+                // Receipt of a PING frame with a length field value other than 8 MUST be treated
+                // as a connection error (Section 5.4.1) of type FRAME_SIZE_ERROR.
+                throw NIOHTTP2Errors.ConnectionError(code: .frameSizeError)
+            }
+            guard streamID == .rootStream else {
+                // PING frames are not associated with any individual stream. If a PING frame is
+                // received with a stream identifier field value other than 0x0, the recipient MUST
+                // respond with a connection error (Section 5.4.1) of type PROTOCOL_ERROR.
+                throw NIOHTTP2Errors.ConnectionError(code: .protocolError)
+            }
+            
+            var tuple: (UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8) = (0,0,0,0,0,0,0,0)
+            withUnsafeMutableBytes(of: &tuple) { ptr -> Void in
+                bytes.readWithUnsafeReadableBytes { bytesPtr -> Int in
+                    ptr.copyBytes(from: bytesPtr[0..<8])
+                    return 8
+                }
+            }
+            
+            payload = .ping(HTTP2PingData(withTuple: tuple))
+            
+        case 7:
+            // GOAWAY frame : RFC 7540 § 6.8
+            guard streamID == .rootStream else {
+                // The GOAWAY frame applies to the connection, not a specific stream. An endpoint
+                // MUST treat a GOAWAY frame with a stream identifier other than 0x0 as a connection
+                // error (Section 5.4.1) of type PROTOCOL_ERROR.
+                throw NIOHTTP2Errors.ConnectionError(code: .protocolError)
+            }
+            
+            guard header.length >= 8 else {
+                // Must have at least 8 bytes of data (last-stream-id plus error-code).
+                throw NIOHTTP2Errors.ConnectionError(code: .frameSizeError)
+            }
+            
+            let raw: UInt32 = bytes.readInteger()!
+            let errcode: UInt32 = bytes.readInteger()!
+            
+            let debugData: ByteBuffer?
+            let extraLen = header.length - 8
+            if extraLen > 0 {
+                debugData = bytes.readSlice(length: extraLen)
+            } else {
+                debugData = nil
+            }
+            
+            payload = .goAway(lastStreamID: HTTP2StreamID(knownID: Int32(raw & ~0x8000_000)),
```

Comment:
We have a lot of silently ignoring the top bit of the stream ID, when I think strictly speaking setting it is supposed to be an error. Should we have an extension on `HTTPStreamID` that adds a throwing initializer that validates the stream ID doesn't have the top bit set, and throws if it does?

Reply:
Good catch. I still think I want an initialiser here, one that expressly indicates that this is a received stream ID and so we should ignore the top bit.